packages/storage/amplify_storage_s3_dart/lib/src/amplify_storage_s3_dart_impl.dart

```diff
@@ -69,19 +111,46 @@ class AmplifyStorageS3Dart extends StoragePluginInterface {
       identityProvider: identityProvider,
     );
 
-    // TODO(HuiSF): create S3Client instance with
-    //  * credentialsProvider
-    //  * _prefixResolver
-    //  * other fields
-    // final credentialsProvider = authProviderRepo
-    //     .getAuthProvider(APIAuthorizationType.iam.authProviderToken);
+    final credentialsProvider = authProviderRepo
+        .getAuthProvider(APIAuthorizationType.iam.authProviderToken);
+
+    if (credentialsProvider == null) {
+      throw const StorageException(
+        'No credential provider found for Storage.',
+        recoverySuggestion:
+            'If you haven\'t already, please add amplify_auth_cognito plugin to your App.',
+      );
+    }
+
+    //`_dependencyManagerOverride` should be available only in unit tests
+    if (_dependencyManagerOverride == null) {
+      _storageS3Service = StorageS3Service(
+        credentialsProvider: credentialsProvider,
+        defaultBucket: _s3pluginConfig.bucket,
+        defaultRegion: _s3pluginConfig.region,
+        prefixResolver: _prefixResolver!,
+        logger: logger,
+      );
+      _dependencyManager.addInstance<StorageS3Service>(_storageS3Service);
+    }
   }
 
   @override
-  StorageListOperation list({
-    required StorageListRequest request,
+  StorageS3ListOperation list({
+    required StorageListRequest<StorageS3ListOptions> request,
   }) {
-    throw UnimplementedError();
+    final s3Service = _dependencyManager.getOrCreate<StorageS3Service>();
```

Comment:
I would make this a getter on the class and use `expect` unless you plan to have code paths that register a builder for it.

```dart
  StorageS3Service get _s3Service => _dependencyManager.expect();
```

Reply:
Ahh yea, wanted to change this line after reading auth impl. but forgot 😅